File: sdks/javascript/src/exceptions.ts
export class EmailConverterError extends Error {
  public statusCode?: number;
  public retryAfter?: number;

  constructor(message: string, statusCode?: number, retryAfter?: number) {
    super(message);
    this.name = 'EmailConverterError';
    this.statusCode = statusCode;
    this.retryAfter = retryAfter;
  }
}

export class AuthenticationError extends EmailConverterError {
  constructor(message = 'Invalid API key') {
    super(message, 401);
    this.name = 'AuthenticationError';
  }
}

export class RateLimitError extends EmailConverterError {
  constructor(message = 'Rate limit exceeded', retryAfter = 60) {
    super(message, 429, retryAfter);
    this.name = 'RateLimitError';
  }
}

export class ValidationError extends EmailConverterError {
  constructor(message: string, statusCode = 400) {
    super(message, statusCode);
    this.name = 'ValidationError';
  }
}

export class NotFoundError extends EmailConverterError {
  constructor(message = 'Resource not found') {
    super(message, 404);
    this.name = 'NotFoundError';
  }
}

export class ServerError extends EmailConverterError {
  constructor(message = 'Server error') {
    super(message, 500);
    this.name = 'ServerError';
  }
}

File: sdks/java/src/main/java/com/emailconverter/sdk/Exceptions.java
package com.emailconverter.sdk;

public class EmailConverterException extends Exception {
    private final int statusCode;

    public EmailConverterException(String message, int statusCode) {
        super(message);
        this.statusCode = statusCode;
    }

    public int getStatusCode() { return statusCode; }
}

public class AuthenticationException extends EmailConverterException {
    public AuthenticationException() { super("Invalid API key", 401); }
}

public class RateLimitException extends EmailConverterException {
    private final int retryAfter;
    public RateLimitException(int retryAfter) { super("Rate limit exceeded", 429); this.retryAfter = retryAfter; }
    public int getRetryAfter() { return retryAfter; }
}
